### PR TITLE
Update to latest Razor Slices

### PR DESCRIPTION
**Pull request by DamianEdwards**
Updates the Platform and Minimal TechEmpower benchmarks to the latest Razor Slices release that uses source generators and supports native AOT. Razor Slices doesn't support rendering to `IBufferWriter<byte>` anymore so had to change `ChunkedBufferWriter` to `ChunkedPipeWriter`.

Removed the `.net7.0` targets as we don't run those anymore and tweaked the version check condition in .csproj files to be compatible with multi-targeting builds.

Also fixed the launch settings in the Platform project so it runs from VS.

Results after the change are within the usual noise margins, i.e. ~566K - ~630K RPS:

image

| db                   | fortunes-platform-baseline | fortunes-platform-update |        |
| -------------------- | -------------------------- | ------------------------ | ------ |
| Max CPU Usage (%)    |                         33 |                       33 |  0.00% |
| Max Cores usage (%)  |                      1,840 |                    1,846 | +0.33% |
| Max Working Set (MB) |                        560 |                      574 | +2.50% |
| Build Time (ms)      |                      1,136 |                    1,146 | +0.88% |
| Start Time (ms)      |                      1,806 |                    1,825 | +1.05% |
| Published Size (KB)  |                    424,584 |                  424,584 |  0.00% |

| application                             | fortunes-platform-baseline | fortunes-platform-update |         |
| --------------------------------------- | -------------------------- | ------------------------ | ------- |
| Max CPU Usage (%)                       |                         72 |                       71 |  -1.39% |
| Max Cores usage (%)                     |                      4,052 |                    3,953 |  -2.44% |
| Max Working Set (MB)                    |                        226 |                      225 |  -0.44% |
| Max Private Memory (MB)                 |                      1,105 |                    1,096 |  -0.81% |
| Build Time (ms)                         |                     11,012 |                    9,099 | -17.37% |
| Start Time (ms)                         |                      9,082 |                    9,411 |  +3.62% |
| Published Size (KB)                     |                    107,395 |                  107,408 |  +0.01% |
| Symbols Size (KB)                       |                         54 |                       56 |  +3.70% |
| .NET Core SDK Version                   |   10.0.100-alpha.1.24615.1 | 10.0.100-alpha.1.24615.1 |         |
| Max CPU Usage (%)                       |                         71 |                       69 |  -3.00% |
| Max Working Set (MB)                    |                        238 |                      235 |  -0.97% |
| Max GC Heap Size (MB)                   |                        101 |                       97 |  -4.71% |
| Size of committed memory by the GC (MB) |                        140 |                      136 |  -3.10% |
| Max Number of Gen 0 GCs / sec           |                      36.00 |                    33.00 |  -8.33% |
| Max Number of Gen 1 GCs / sec           |                       2.00 |                     2.00 |   0.00% |
| Max Number of Gen 2 GCs / sec           |                       2.00 |                     1.00 | -50.00% |
| Max Gen 0 GC Budget (MB)                |                         73 |                       71 |  -2.74% |
| Max Time in GC (%)                      |                      18.00 |                     7.00 | -61.11% |
| Max Gen 0 Size (B)                      |                  7,691,104 |                1,129,032 | -85.32% |
| Max Gen 1 Size (B)                      |                 25,435,552 |                9,425,704 | -62.94% |
| Max Gen 2 Size (B)                      |                 19,904,576 |               20,596,264 |  +3.48% |
| Max LOH Size (B)                        |                    192,448 |                  192,448 |   0.00% |
| Max POH Size (B)                        |                  4,214,568 |                4,235,168 |  +0.49% |
| Max Allocation Rate (B/sec)             |              1,994,380,112 |            2,006,939,776 |  +0.63% |
| Max GC Heap Fragmentation (%)           |                     5,334% |                   3,429% | -35.71% |
| # of Assemblies Loaded                  |                         93 |                       89 |  -4.30% |
| Max Exceptions (#/s)                    |                        964 |                      952 |  -1.24% |
| Max Lock Contention (#/s)               |                      5,959 |                    6,104 |  +2.43% |
| Max ThreadPool Threads Count            |                         66 |                       73 | +10.61% |
| Max ThreadPool Queue Length             |                         68 |                       67 |  -1.47% |
| Max ThreadPool Items (#/s)              |                  1,636,624 |                1,645,142 |  +0.52% |
| Max Active Timers                       |                         44 |                       44 |   0.00% |
| IL Jitted (B)                           |                    644,635 |                  647,966 |  +0.52% |
| Methods Jitted                          |                      7,785 |                    7,863 |  +1.00% |
| Load Working Set - P90 (MB)             |                        220 |                      225 |  +2.27% |
| Load CPU Usage - P90 (%)                |                         63 |                       65 |  +3.17% |

| load                    | fortunes-platform-baseline | fortunes-platform-update |         |
| ----------------------- | -------------------------- | ------------------------ | ------- |
| Max CPU Usage (%)       |                         25 |                       23 |  -8.00% |
| Max Cores usage (%)     |                      1,375 |                    1,277 |  -7.13% |
| Max Working Set (MB)    |                         43 |                       43 |   0.00% |
| Max Private Memory (MB) |                        363 |                      363 |   0.00% |
| Build Time (ms)         |                      7,973 |                        0 |         |
| Start Time (ms)         |                          1 |                        1 |   0.00% |
| Published Size (KB)     |                     72,308 |                        0 |         |
| Symbols Size (KB)       |                          0 |                        0 |         |
| .NET Core SDK Version   |                    8.0.404 |                        0 |         |
| First Request (ms)      |                        128 |                      151 | +17.97% |
| Requests/sec            |                    620,856 |                  622,143 |  +0.21% |
| Requests                |                  9,374,749 |                9,394,489 |  +0.21% |
| Mean latency (ms)       |                       0.84 |                     0.84 |  -0.31% |
| Max latency (ms)        |                       6.91 |                     6.87 |  -0.58% |
| Bad responses           |                          0 |                        0 |         |
| Socket errors           |                          0 |                        0 |         |
| Read throughput (MB/s)  |                     814.72 |                   816.41 |  +0.21% |
| Latency 50th (ms)       |                       0.76 |                     0.77 |  +0.53% |
| Latency 75th (ms)       |                       1.02 |                     1.02 |   0.00% |
| Latency 90th (ms)       |                       1.34 |                     1.34 |   0.00% |
| Latency 99th (ms)       |                       2.52 |                     2.43 |  -3.57% |

### CODE REVIEW

File: src/BenchmarksApps/Grpc/GrpcHttpApiServer/Server/Program.cs [hunk 1]
```diff
@@ -84,7 +85,10 @@ public static IHostBuilder CreateHostBuilder(string[] args)
                     if (Enum.TryParse<LogLevel>(config["LogLevel"], out var logLevel))
                     {
                         Console.WriteLine($"Console Logging enabled with level '{logLevel}'");
-                        loggerFactory.AddConsole(o => o.TimestampFormat = "ss.ffff ").SetMinimumLevel(logLevel);
+                        loggerFactory
+                            .AddConsole()
+                            .AddConsoleFormatter<ConsoleFormatter, ConsoleFormatterOptions>(o => o.TimestampFormat = "ss.ffff")
```
Comment (DamianEdwards) [inline src/BenchmarksApps/Grpc/GrpcHttpApiServer/Server/Program.cs:90]: This change was done to resolve an obsolete warning.

File: src/BenchmarksApps/Grpc/GrpcHttpApiServer/Server/Server.csproj [hunk 2]
```diff
@@ -10,7 +10,7 @@
 
   <ItemGroup>
     <PackageReference Include="Grpc.AspNetCore" Version="2.27.0" />
-    <PackageReference Include="Microsoft.AspNetCore.Grpc.HttpApi" Version="0.1.0-alpha.20121.1" />
+    <PackageReference Include="Microsoft.AspNetCore.Grpc.HttpApi" Version="0.1.0-alpha.20179.2" />
```
Comment (DamianEdwards) [inline src/BenchmarksApps/Grpc/GrpcHttpApiServer/Server/Server.csproj:13]: This was done to quiet a NuGet warning about the previous stated version being unresolvable.

File: src/Benchmarks/Benchmarks.csproj [hunk 3]
```diff
@@ -62,7 +62,7 @@
     <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="$(MicrosoftEntityFrameworkCoreSqlServerVersion90)" />
     <PackageReference Include="Microsoft.Data.SqlClient" Version="$(MicrosoftDataSqlClientVersion90)" />
     <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="$(MicrosoftEntityFrameworkCoreSqliteVersion90)" />
-    <PackageReference Include="Microsoft.AspNetCore.Server.IntegrationTesting.IIS" Version="$(MicrosoftAspNetCoreAppPackageVersion)" />
+    <PackageReference Include="Microsoft.AspNetCore.Server.IntegrationTesting.IIS" Version="$(MicrosoftIISTestingVersion90)" />
```
Comment (DamianEdwards) [inline src/Benchmarks/Benchmarks.csproj:65]: Seems this was a typo/copy-pasta that was causing a NuGet warning as the versions didn't line up.

File: src/BenchmarksApps/TechEmpower/BlazorSSR/BlazorSSR.csproj [hunk 4]
```diff
@@ -14,7 +14,7 @@
     <PackageReference Include="Npgsql" Version="$(NpgsqlVersion80)" />
   </ItemGroup>
 
-  <ItemGroup Condition="$([MSBuild]::VersionGreaterThanOrEquals($(TargetFrameworkVersion), '9.0'))">
+  <ItemGroup Condition="$([MSBuild]::IsTargetFrameworkCompatible($(TargetFramework), 'net9.0'))">
```
Comment (DamianEdwards) [inline src/BenchmarksApps/TechEmpower/BlazorSSR/BlazorSSR.csproj:17]: This fixes a build error that occurs when building the project without specifying a target framework, e.g. when building from VS :smile:

File: src/BenchmarksApps/TechEmpower/Minimal/Program.cs [hunk 5]
```diff
@@ -37,12 +38,11 @@
 
 app.MapGet("/db/result", async (Db db) => Results.Json(await db.LoadSingleQueryRow()));
 
-var createFortunesTemplate = RazorSlice.ResolveSliceFactory<List<Fortune>>("/Templates/Fortunes.cshtml");
 var htmlEncoder = CreateHtmlEncoder();
 
 app.MapGet("/fortunes", async (HttpContext context, Db db) => {
     var fortunes = await db.LoadFortunesRows();
-    var template = (RazorSliceHttpResult<List<Fortune>>)createFortunesTemplate(fortunes);
+    var template = (RazorSliceHttpResult<List<Fortune>>)Fortunes.Create(fortunes);
```
Comment (DamianEdwards) [inline src/BenchmarksApps/TechEmpower/Minimal/Program.cs:45]: Razor Slices generates these template factory classes now.

File: src/BenchmarksApps/TechEmpower/PlatformBenchmarks/BenchmarkApplication.Fortunes.cs [hunk 6]
```diff
@@ -1,81 +1,81 @@
...
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
 using System.IO.Pipelines;
 using System.Runtime.CompilerServices;
 using System.Threading.Tasks;
 using RazorSlices;
 
-namespace PlatformBenchmarks
+namespace PlatformBenchmarks;
```
Comment (DamianEdwards) [inline src/BenchmarksApps/TechEmpower/PlatformBenchmarks/BenchmarkApplication.Fortunes.cs:10]: Changed to file-scoped namespace declaration on this file.

File: src/BenchmarksApps/TechEmpower/PlatformBenchmarks/Program.cs [hunk 7]
```diff
@@ -85,6 +86,7 @@ public static IWebHost BuildWebHost(string[] args)
 
             var hostBuilder = new WebHostBuilder()
                 .UseBenchmarksConfiguration(config)
+                //.ConfigureLogging(logging => logging.AddConsole())
```
Comment (DamianEdwards) [inline src/BenchmarksApps/TechEmpower/PlatformBenchmarks/Program.cs:89]: Will fix this

File: src/BenchmarksApps/TechEmpower/PlatformBenchmarks/PlatformBenchmarks.binlog
Comment (BrennanConroy) [inline src/BenchmarksApps/TechEmpower/PlatformBenchmarks/PlatformBenchmarks.binlog:1]: Remove

File: src/BenchmarksApps/TechEmpower/PlatformBenchmarks/BenchmarkApplication.cs [hunk 8]
```diff
@@ -10,216 +10,216 @@
...
+            writer.Reset();
+            return true;
         }
+    }
 
 #if DATABASE
-        private readonly static SliceFactory<List<FortuneUtf8>> FortunesTemplateFactory = RazorSlice.ResolveSliceFactory<List<FortuneUtf8>>("/Templates/FortunesUtf8.cshtml");
-        private readonly static SliceFactory<List<FortuneUtf16>> FortunesDapperTemplateFactory = RazorSlice.ResolveSliceFactory<List<FortuneUtf16>>("/Templates/FortunesUtf16.cshtml");
-        private readonly static SliceFactory<List<FortuneEf>> FortunesEfTemplateFactory = RazorSlice.ResolveSliceFactory<List<FortuneEf>>("/Templates/FortunesEf.cshtml");
+    //private readonly static Func<List<FortuneUtf8>, RazorSlice> FortunesTemplateFactory = FortunesUtf8.Create(;
```
Comment (BrennanConroy) [inline src/BenchmarksApps/TechEmpower/PlatformBenchmarks/BenchmarkApplication.cs:61]: Delete?

File: src/BenchmarksApps/TechEmpower/PlatformBenchmarks/ChunkedPipeWriter.cs [hunk 9]
```diff
@@ -2,18 +2,21 @@
...
+using System.IO.Pipelines;
 using System.Numerics;
 using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
 
 namespace PlatformBenchmarks;
 
-internal sealed class ChunkedBufferWriter<TWriter> : IBufferWriter<byte> where TWriter : IBufferWriter<byte>
+internal sealed class ChunkedPipeWriter : PipeWriter
```
Comment (BrennanConroy) [inline src/BenchmarksApps/TechEmpower/PlatformBenchmarks/ChunkedPipeWriter.cs:13]: nit: implement `UnflushedBytes`

**Review by sebastienros — APPROVED** [review]

**Review by BrennanConroy — APPROVED** [review]

**Pull request merged**